### PR TITLE
Added initial Drone pipeline

### PR DESCRIPTION
**Pull request by robertstettner**
This is using Drone to accomplish the CI part for the project.

### CODE REVIEW

**Comment by aviflax** [comment]
Thanks Robert!

Based on your comments in #26:

> …it is clear to me that GitHub Actions for CI/CD is not easy, and looks complex.

and the existence of this PR, I’m inferring that you are thinking that Drone might be a better fit for this project than GitHub Actions. Is that right?

If so, I’m happy to discuss the relative pros and cons of Drone and Actions, but first I’d want to have a better idea of how Drone would be viable for an open-source project such as this. I assume you’re not suggesting that we (Funding Circle (FC)) run all the automation workflows/pipelines for this project within our private Drone installation, because those runs wouldn’t be accessible to non-FC contributors. So are you thinking that we’d use [Drone Cloud](https://cloud.drone.io/)?

If so, I’m definitely open to that, but I am a little concerned to see that [their installation guide for GitHub projects](https://docs.drone.io/server/provider/github/) directs people to create and maintain GitHub OAuth applications… this is somewhat outmoded; GitHub recommends their “Apps” feature rather than OAuth apps for [a long list of reasons](https://developer.github.com/apps/migrating-oauth-apps-to-github-apps/#reasons-for-switching-to-github-apps) including some security reasons. So I’d be concerned about the maintenance burden and the security risks of creating, maintaining, and using an OAuth app. More broadly though, this makes me concerned about the degree of investment being made into Drone Cloud. GitHub has been encouraging their entire ecosystem to transition [since 2017](https://github.blog/2018-09-25-migrating-to-github-apps-code-climate-shares-their-story/)… OTOH, it’s non-trivial work, so I totally understand sometimes things just take awhile (e.g. Jenkins [migrated just last month](https://www.jenkins.io/blog/2020/04/16/github-app-authentication/)) — but in this case it’s an unknown, something we’d need to look into.

What do you think?

File: .drone.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  test:
+    image: clojure:openjdk-11-tools-deps-1.10.1.502-slim-buster
+    commands: clj -A:dev:test
+    <<: *when-feature-branch
+
+  push-cache:
+    image: robertstettner/drone-s3cache:2
+    bucket: 131063299351-drone-cache
+    region: eu-west-1
+    folders: .cpcache
```
Comment (aviflax) [inline .drone.yml:18]:
```suggestion
    folders:
      - .cpcache
      - .gitlibs
```

File: .drone.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  pull-cache:
+    image: robertstettner/drone-s3cache:2
+    bucket: 131063299351-drone-cache
+    region: eu-west-1
+    mode: pull
+    <<: *when-feature-branch
+
+  test:
+    image: clojure:openjdk-11-tools-deps-1.10.1.502-slim-buster
+    commands: clj -A:dev:test
```
Comment (aviflax) [inline .drone.yml:11]:
```suggestion
    commands: bin/kaocha
```

**Comment by robertstettner** [comment]
> I’m inferring that you are thinking that Drone might be a better fit for this project than GitHub Actions. Is that right?

From what I see and know of GitHub Actions, it is definitely easier to manage. I hope you can see that in the syntax and ease of use.

> I assume you’re not suggesting that we (Funding Circle (FC)) run all the automation workflows/pipelines for this project within our private Drone installation, because those runs wouldn’t be accessible to non-FC contributors.

It is possible to use our self-served Drone instance to run pipelines for our open-source projects. When developers fork the repo, and open a pull request back, Drone has a PR event that can be used for conditions in the pipeline. The question is should we go down that route for now, in particular for open-source projects. As with everything I think we should try it, and evaluate. If you think that this isn't time for this particular project. I will close this PR.

> So I’d be concerned about the maintenance burden and the security risks of creating, maintaining, and using an OAuth app.

Your concerns are valid, and this might be the defining problem especially when using Drone for open-source projects. I think we should evaluate this.

**Comment by robertstettner** [comment]
> Also: why exclude master?

Just a temporary measure, I didn't want to interfere with anything else running on `master` branch.

**Comment by aviflax** [comment]
> From what I see and know of GitHub Actions, it is definitely easier to manage.

OK, cool, that makes your perspective crystal clear — thanks!

> I hope you can see that in the syntax and ease of use.

I don’t really know what you mean by “ease of use” so I’d be interested in having you break that down.

As for the syntax, I was thinking about this yesterday, as I wrote in #26:

> I can see how this specific job might give that impression [that GitHub Actions is complex] at a glance, because there are a lot of lines related to caching.

I think the job definition I added in that PR was a poor representative of the GHA definition language, which I think is quite good, actually.

So in order to test that hypothesis, I spent a few minutes seeing how well I could tighten up the definition of that job, to make it simpler, clearer, and way more concise. You can see the diff for this in fcfe88101fd5, but the end result is short enough that I’ll just paste it here:

```yaml
name: Test
on: push
jobs:
  examples:
    runs-on: ubuntu-18.04
    container: clojure:openjdk-11-tools-deps-1.10.1.502-slim-buster
    steps:
      - uses: actions/checkout@v2
      - uses: actions/cache@bac1a40c # We need a not-yet-released feature: https://git.io/JfBDh
        env: { cache-version: v3 }
        with:
          path: |
            .cpcache
            /root/.m2
            /root/.gitlibs
          key: ${{ runner.os }}-${{ github.workflow }}-${{ env.cache-version }}-${{ hashFiles('deps.edn') }}
      - run: bin/kaocha examples
```

I was quite happy with that; I think it’s simple, clear, and concise.

Encouraged by that result, I thought I’d then try to explore the GHA definition language a bit more, to go a bit deeper, to try to find out how “powerful” I could make the job definition while staying sufficiently simple, clear, and concise.

So I added some more commits on top of fcfe88101fd5:

* In 22cb2b1857e1 I added a matrix strategy to the job, so it runs multiple times with different combinations of JVM versions and test suites.
* In f81b683c6e6 I added a version of the job that runs on MacOS
* In d9ab8c183f2 I merged the two jobs into a single job that runs on both Linux and MacOS

In the end, after working through all of this, I ended up with a job definition that I think is really quite good:

```yaml
name: Test
on: push
jobs:
  Kaocha:
    strategy:
      matrix:
        suite: [examples, integration, properties]
        os: [macos-10.15, ubuntu-18.04]
        java: [11, 13, 14]
    runs-on: ${{ matrix.os }}
    steps:
      - uses: actions/checkout@v2
      - uses: actions/cache@bac1a40c # We need a not-yet-released feature: https://git.io/JfBDh
        env: { cache-version: v4 }
        with:
          path: |
            .cpcache
            ~/.m2
            ~/.gitlibs
          key: ${{ runner.os }}-${{ github.workflow }}-${{ env.cache-version }}-${{ hashFiles('deps.edn') }}
      - uses: actions/setup-java@v1
        with:
          java-version: ${{ matrix.java }}
          java-package: jre
      - uses: DeLaGuardo/setup-clojure@2.0
        with: { tools-deps: '1.10.1.502' }
      - run: bin/kaocha ${{ matrix.suite }}
```

That one job definition yields, when the workflow is triggered, 18 job executions; one for each combination of OS, JVM, and test suite. They all succeed and they all run quickly due to effective caching. For an example see [this workflow run](https://github.com/FundingCircle/md2c8e/actions/runs/105927409). Actually, I think this is sufficiently cool that it deserves a screenshot:

image

I’d love to hear what you think of all this, having now seen a better and more substantial example. 

> It is possible to use our self-served Drone instance to run pipelines for our open-source projects. When developers fork the repo, and open a pull request back, Drone has a PR event that can be used for conditions in the pipeline.

I certainly see how it would be possible, in a technical sense, to make the pipelines run when desired. My big concern with this scenario would be whether external contributors would be able to access the Drone Web GUI so they could dig into logs, etc — to interrogate and scrutinize the results of a run, whether for debugging, or optimization, etc.

I suspect that enabling external access to our instance of the Drone Web GUI would be not worth the effort and risk involved, from a security perspective. Neither, I suspect, would it be cost-effective for us to configure a separate instance specifically for external third parties.

(Another cost factor I’d be concerned about is the resource utilization, and possibly contention, that might be entailed by continuing to run our automations for our internal projects and our open-source projects in the same pool of resources. I’ve seen some of the jobs for some of our open-source projects be queued behind jobs for our internal projects, so the converse must be true as well. And since our new GitHub plan will have some degree of GHA resources bundled with it, it seems to me it might be smart to utilize those resources as a way to reduce the cost impact of our open-source automations — both monetary costs and opportunity costs.)

> The question is should we go down that route for now, in particular for open-source projects. As with everything I think we should try it, and evaluate. If you think that this isn't time for this particular project. I will close this PR.

I appreciate your chill on this!

I _do_ think that this isn’t the time to try Drone for this particular project. And perhaps also, more broadly, for all or most of our open-source projects.

That said, thank you for suggesting it and demonstrating it. I think it’s hugely valuable to explore possibilities and consider alternatives. Cheers!

**Pull request closed without merging**